go/flexver/main.go

```diff
@@ -0,0 +1,264 @@
+/*
+ * To the extent possible under law, the author has dedicated all copyright
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * See <http://creativecommons.org/publicdomain/zero/1.0/>
+ */
+
+// Package flexver implements FlexVer, a SemVer-compatible intuitive comparator for free-form versioning strings as
+// seen in the wild. It's designed to sort versions like people do, rather than attempting to force
+// conformance to a rigid and limited standard. As such, it imposes no restrictions. Comparing two
+// versions with differing formats will likely produce nonsensical results (garbage in, garbage out),
+// but best effort is made to correct for basic structural changes, and versions of differing length
+// will be parsed in a logical fashion.
+//
+// See the specification at https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md
+package flexver
+
+import (
+	"errors"
+	"sort"
+	"unicode/utf8"
+)
+
+type componentType uint8
+
+const (
+	componentTypeTextual componentType = iota
+	componentTypeNumeric
+	componentTypePreRelease
+	// componentTypeAppendix (unused; appendix components removed in decompose)
+	componentTypeNull
+)
+
+var nullComponent = component{t: componentTypeNull, value: nil}
+
+type component struct {
+	t     componentType
+	value []rune
+}
+
+// compare determines if this component is less than, equal to or greater than the given component.
+// A negative, zero or positive integer respectively indicates this result.
+func (compA component) compare(compB component) int32 {
+	// If both components are numeric, compare numerically (codepoint-wise)
+	if compA.t == componentTypeNumeric && compB.t == componentTypeNumeric {
+		var i, j int
+		// Ignore leading zeroes
+		for i < len(compA.value) && compA.value[i] == '0' {
+			i++
+		}
+		for j < len(compB.value) && compB.value[j] == '0' {
+			j++
+		}
+		aLen := len(compA.value) - i
+		if aLen != len(compB.value)-j {
+			// Lengths differ; compare by length
+			return int32(aLen - (len(compB.value) - j))
```

Comment:
Shouldn't this call `signum` (or equivalent) instead of casting to `int32`? While highly unlikely, this could result in an incorrect answer in the event of overflow.

Reply:
Go has no equivalent `signum` built-in function (hm yes "batteries-included" standard library). That makes sense though, I'll add my own :P